### DISCUSSION TITLE
Is it possible to execute a command using multithread?

### DISCUSSION THREAD

**Discussion by citystrawman** [Tyrrrz/CliWrap#244] [Q&A] [answered]
Hi, I want to execute a command simultaneously. The single command looks as follows:

```
bashPath = "C:\\Windows\\System32\\bash.exe";

String command = $"export FREESURFER_HOME={sourceScriptDirWsl};source {sourceScriptDirWsl}/SetUpFreeSurfer.sh;{reconAllPathWsl} -s {outputName} -i {inputPathWsl} -all";

var result = await Cli.Wrap(bashPath)
    .WithArguments(args => args
    .Add("-c")
    .Add(command))
    .ExecuteBufferedAsync(cts.Token);
```

to explain this, it uses `bash -c` to run a command to process a file within WSL. what I want to do is that I have multiple files to process, and because this command is single-core based, I would want to use multithread to execute it simultaneously. 

May I know is it possible to execute a command using multithread, and how?

I appreciate if someone could help me. Thanks a lot!

**Comment by Tyrrrz** [comment]
Just don't `await` directly and use `Task.WhenAll(...)` instead

**Reply by citystrawman** [reply]
suppose I know the number of logical cores of the computer, then is it the most efficient way to devide the whole work into several parts, where each part have number of logical cores of jobs, and then apply Task.WhenAll for each part?

**Reply by Tyrrrz** [reply]
That's what some people do but I'm not sure if that's the most efficient practice. You'd have to benchmark it.

**Reply by citystrawman** [reply]
So it is not necessary to devide the whole work and just use WhenAll is OK?

**Reply by Tyrrrz** [reply, marked as answer]
Yes
